### PR TITLE
Add streams group member

### PR DESCRIPTION
**Pull request by cadonna**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java [hunk 1]
```diff
@@ -0,0 +1,674 @@
...
+        private int memberEpoch = 0;
+        private int previousMemberEpoch = -1;
+        private MemberState state = MemberState.STABLE;
+        private String instanceId = null;
+        private String rackId = null;
+        private int rebalanceTimeoutMs = -1;
+        private String clientId = "";
+        private String clientHost = "";
+        private Set<String> subscribedTopicNames = Collections.emptySet();
+        private String subscribedTopicRegex = "";
```
Comment (lucasbru) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java]: do we still need this?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java [hunk 2]
```diff
@@ -0,0 +1,674 @@
...
+        private final String memberId;
+        private int memberEpoch = 0;
+        private int previousMemberEpoch = -1;
+        private MemberState state = MemberState.STABLE;
+        private String instanceId = null;
+        private String rackId = null;
+        private int rebalanceTimeoutMs = -1;
+        private String clientId = "";
+        private String clientHost = "";
+        private Set<String> subscribedTopicNames = Collections.emptySet();
```
Comment (lucasbru) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java]: do we still need this?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+        result = 31 * result + Objects.hashCode(serverAssignorName);
+        result = 31 * result + Objects.hashCode(assignedPartitions);
+        result = 31 * result + Objects.hashCode(partitionsPendingRevocation);
+        result = 31 * result + Objects.hashCode(classicMemberMetadata);
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "ConsumerGroupMember(" +
```
Comment (lucasbru) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java]: StreamsGroupMember?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java (hunk 2 above)
Comment (cadonna) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java]: No, I will remove it

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java (hunk 1 above)
Comment (cadonna) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java]: Same here

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java (hunk 3 above)
Comment (cadonna) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroupMember.java]: Fixed.

**Review by lucasbru — APPROVED** [review]
lgtm, thnaks

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java [hunk 4]
```diff
@@ -0,0 +1,1211 @@
...
+
+    /**
+     * Compute the subscription type of the consumer group.
+     *
+     * @param subscribedTopicNames      A map of topic names to the count of members subscribed to each topic.
+     *
+     * @return {@link org.apache.kafka.coordinator.group.assignor.SubscriptionType#HOMOGENEOUS} if all members are subscribed to exactly the same topics;
+     *         otherwise, {@link org.apache.kafka.coordinator.group.assignor.SubscriptionType#HETEROGENEOUS}.
+     */
+    public static SubscriptionType subscriptionType(
```
Comment (lucasbru) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:856]: I don't think streams group have a notion of individual subscript. I think this shouldn't be here.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java [hunk 5]
```diff
@@ -0,0 +1,1211 @@
...
+     */
+    @Override
+    public boolean isSubscribedToTopic(String topic) {
+        return subscribedTopicNames.containsKey(topic);
+    }
+
+    /**
+     * @return The group's subscription type.
+     */
+    public SubscriptionType subscriptionType() {
```
Comment (lucasbru) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java]: remove

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java [hunk 6]
```diff
@@ -0,0 +1,1211 @@
...
+
+    /**
+     * The number of members supporting each assignor name.
+     */
+    private final TimelineHashMap<String, Integer> assignors;
+
+    /**
+     * The number of subscribers per topic.
+     */
+    private final TimelineHashMap<String, Integer> subscribedTopicNames;
```
Comment (lucasbru) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java]: There is a lot of information about subscription here. As I understand it, streams clients do not susbcribe to topics, instead we initialize a topology. Why do we have all this subscription stuff here?

Comment (cadonna) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java]: Yeah, I was not sure if we needed the fields and methods. It was not clear to me if we plan to get this information from the initialize RPC and inject it into the Streams group. For now, I removed the obvious ones, but did not check if I can remove more. I think for a PoC that is fine.

**Pull request merged**